cores/esp8266/StreamDev.h

```diff
@@ -0,0 +1,247 @@
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#ifndef __STREAMDEV_H
+#define __STREAMDEV_H
+
+#include <esp_priv.h>
+#include <StreamString.h>
+
+///////////////////////////////////////////////
+// /dev/null
+// - black hole as output, swallow everything, availableForWrite = infinite
+// - black hole as input, nothing to read, available = 0
+
+class StreamNull: public Stream
+{
+public:
+
+    // Print
+    virtual size_t write(uint8_t) override
+    {
+        return 1;
+    }
+
+    virtual size_t write(const uint8_t* buffer, size_t size) override
+    {
+        (void)buffer;
+        (void)size;
+        return size;
+    }
+
+    virtual int availableForWrite() override
+    {
+        return 32767;
+    }
+
+    // Stream
+    virtual int available() override
+    {
+        return 0;
+    }
+
+    virtual int read() override
+    {
+        return -1;
+    }
+
+    virtual int peek() override
+    {
+        return -1;
+    }
+
+    virtual size_t readBytes(char* buffer, size_t len) override
+    {
+        (void)buffer;
+        (void)len;
+        return 0;
+    }
+
+    virtual int read(uint8_t* buffer, size_t len) override
+    {
+        (void)buffer;
+        (void)len;
+        return 0;
+    }
+
+    virtual bool outputTimeoutPossible() override
+    {
+        return false;
+    }
+
+    virtual bool inputTimeoutPossible() override
+    {
+        return false;
+    }
+
+    virtual ssize_t streamSize() override
+    {
+        return 0;
+    }
+};
+
+///////////////////////////////////////////////
+// /dev/zero
+// - black hole as output, swallow everything, availableForWrite = infinite
+// - big bang as input, gives infinity to read, available = infinite
+
+class StreamZero: public StreamNull
+{
+protected:
+
+    char _zero;
+
+public:
+
+    StreamZero(char zero = 0): _zero(zero) { }
+
+    // Stream
+    virtual int available() override
+    {
+        return 32767;
+    }
+
+    virtual int read() override
+    {
+        return _zero;
+    }
+
+    virtual int peek() override
+    {
+        return _zero;
+    }
+
+    virtual size_t readBytes(char* buffer, size_t len) override
+    {
+        memset(buffer, _zero, len);
+        return len;
+    }
+
+    virtual int read(uint8_t* buffer, size_t len) override
+    {
+        memset((char*)buffer, _zero, len);
+        return len;
+    }
+
+    virtual ssize_t streamSize() override
+    {
+        return 32767;
+    }
+};
+
+///////////////////////////////////////////////
+// static buffer (in flash or ram)
+// - black hole as output, swallow everything, availableForWrite = infinite
+// - Stream buffer out as input, resettable
+
+class StreamPtr: public StreamNull
```

Comment:
Why take a String as input and not a const char * directly?
What is the difference between this and StringStream or S2Stream?
I would make the array element type a template parameter:
```cpp
template <ElementType>
class StreamArray : public StreamNull
{
...
  const ElementType * _buffer;
  StreamArray(const ElementType * array) : _buffer(array) ...
};
```
Also, maybe rename it to StreamConstPtr to emphasize that the array data can't be touched (Ptr to Const, read backwards), but I don't feel strongly about the rename, up to you.

ElementType would be a pod type or contiguous non-polymorphic struct, and must not be some complex type.
Also, cast the _buffer to uint8_t * before passing to the "normal" underlying Stream methods, and pass size * sizeof(ElementType) instead of size directly.
Also, this is likely where you want to handle arrays that are stored in flash (is that already the case in this implementation? I can't tell)

With that you should be able to stream chars, ints, doubles, simple structs, whatever.
As for String, let the StringStream implementation handle that case, including the case where you don't want to make a copy. Handling that here adds to the confusion of the String vs. Stream apocalypse.

Reply:
> What is the difference between StreamPtr and StringStream or S2Stream?

There is no allocation in `StreamPtr::`. It's an interface that handles a ***pointer*** to make a `PROGMEM` or a `char*` directly usable as a Stream as input regardless whether data is byte readable.
The *"`Ptr`"* in its name makes it clear that the pointed object must not be modified while beeing used (speaking of `String::c_str()` as pointed object).

> StreamConstPtr

OK with this name

> Why take a String as input and not a const char * directly?

Both and others are taken. 
Taking `String` avoids users to give `myString.c_str()` as parameter which is just what we usually do for ease of reading and coding.

> template <ElementType>

There are special cases to consider. `PROGMEM`/`F()` are *not* byte-accessible. `char*` can be optimized but not when it's a `PROGMEM`.
Please check the constructors about them.

> With that you should be able to stream chars, ints, doubles, simple structs, whatever.

The purpose of this class is to make no allocation.
I guess that you are thinking of `S2Stream::`.

> Handling `String` here adds to the confusion of the String vs. Stream apocalypse

Well, it has been made clear that `String` is not and should never be a `Stream`.
Given the name `StreamConstPtr` of this class I don't think confusion will ever be made.
This class is lighter that `S2Stream::` so why not letting users use a `String` with it ?